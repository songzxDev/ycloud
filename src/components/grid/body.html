<div class="y-grid-body">
  <table cellspacing="0" cellpadding="0" border="0" data-bind="attr:{id:domId}">
    <colgroup data-bind="foreach:{data: columns,as: 'col'}">
      <col data-bind="visible: !col.hidden, attr:{width:col.width}">
    </colgroup>
    <tbody data-bind="foreach: {data: rows, as: 'row'}">

      <tr data-bind="css:{'y-grid-row-hover':row._hover,'y-grid-row-selected':row._selected},event:{click:$parent.handleClick,mouseenter:$parent.handleMouseIn,mouseleave:$parent.handleMouseOut}">
        <!-- ko foreach:{data: $parent.columns, as: 'col'} -->
        <td>
          <y-grid-td params="isDataTable:$parentContext.$parent.isDataTable,row: row, col: col, rowIndex: $parentContext.$index(),colIndex:$index()"></y-grid-td>
        </td>
        <!-- /ko -->
      </tr>
      <!-- ko if: $parent.expand && row._expand() -->
        <tr>
          <td data-bind="attr:{colspan: $parentContext.$parent.columns.length}">
            <div class="y-cell" data-bind="component: $parentContext.$parent.columns[0].expandCompFn(row)">
            </div>
          </td>
        </tr>
      <!-- /ko -->

    </tbody>
    <tbody class="y-grid-empty">
      <tr>
        <td data-bind="attr:{colspan: columns.length},style:{height: noDataHeight}">
          暂无数据
        </td>
      </tr>
    </tbody>
  </table>
</div>
