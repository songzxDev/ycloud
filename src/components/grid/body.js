import template from './body.html'
import ko from 'knockout'
import td from './td'
// 判断点击事件是否触发行选中
function isTriggerRowSelect (evt) {
  // 如果点击的是input则不触发
  if (evt.target.tagName === 'INPUT') {
    return false
  }
  if (evt.target.tagName === 'A') {
    return false
  }
  if (evt.target.classList.contains('y-grid-operation')) {
    return false
  }
  return true
}
const PREFIX = 'y-'
ko.components.register(PREFIX + td.name, {
  viewModel: td.init,
  template: td.template
})
function init (params) {
  // 为跨页选择做准备
  // let cacheData = []
  this.isDataTable = params.isDataTable
  this.setCellWidth = function () { return '200' }
  this.expand = params.expand
  this.columns = params.columns
  this.rows = ko.computed(() => {
    params.rows().forEach((row) => {
      row._hover = ko.observable(false)
      row._disabled = ko.observable(false)
      row._expand = ko.observable(false)
      if (this.isDataTable) {
        // 如果是dataTable设置成计算属性
        row._selected = ko.pureComputed({
          read: function () {
            return row.selected()
          },
          write: function (val) {
            if (val) {
              row.parent.addRowSelect(row)
            } else {
              row.parent.setRowUnSelect(row)
            }
          }
        })
      } else {
        row._selected = ko.observable(false)
      }
      // todo:行是否选中要和cacheData里的数据进行合并
    })
    return params.rows()
  })
  // 暂无数据区域的高度
  this.noDataHeight = ko.computed(() => {
    if (params.maxheight === 'auto') {
      return 'auto'
    } else {
      return params.maxheight.replace('px', '') - 150 + 'px'
    }
  })
  this.handleMouseIn = (row) => {
    if (row._hover()) return
    row._hover(true)
  }
  this.handleMouseOut = (row) => {
    row._hover(false)
  }
  this.handleClick = (row, evt) => {
    // 对于一些输入框不选中
    if (!isTriggerRowSelect(evt)) {
      return true
    }
    // todo:需要将选中的数据放入缓存当中或从缓存中剔除
    let _selected = !row._selected()
    row._selected(_selected)
    // 触发行选中事件
    params.onRowSelect && params.onRowSelect(row)
    return true
  }
}

export default {
  name: 'grid-body',
  init,
  template
}
