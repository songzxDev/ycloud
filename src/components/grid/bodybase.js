import ko from 'knockout'
import Base from '../../core/base'
import util from './gridutil'
import getLang from '../../i18n'
class Body extends Base {
  initialize (params) {
    this.lazy = params.lazy
    this.i18nNoData = getLang('暂无数据')
    this.hasSummaryRow = params.hasSummaryRow || false
    // 用在子表一对多的场景之中
    this.hasChildGrid = params.hasChildGrid || false
    this.isDataTable = params.isDataTable
    this.caculateColumns = params.caculateColumns
    this.expand = params.expand
    this.columns = params.columns
    this.rows = params.rows
    this.domId = params.domId
    this.onRowSelect = params.onRowSelect
    this.defaultExpand = params.defaultExpand || false
    this.isNoDataCustomTpl = params.isNoDataCustomTpl || false
    this.lockhead = params.lockhead
    this.tdstyle = params.tdstyle
    this.isSeparate = params.isSeparate
    this.forbitRowSelect = params.forbitRowSelect || function () {}
    this.forbitRowSelectFn = params.forbitRowSelectFn || function () { return false }
    this.isShowLoading = params.isShowLoading
    this.transform = params.transform
    this.headHeight = params.headHeight
    this.verticalAlign = params.verticalAlign
    this.lockColumnHeight = params.lockColumnHeight
    this.isLockLeft = params.isLockLeft || false
    this.tableWidth = params.isLockLeft ? function () {
      return '100%'
    } : params.tableWidth
    this.computedLockColumnHeight = ko.observable({})
    this.computedLockColumnHeight.subscribe((val) => {
      var length = this.rows.peek().length
      var count = 0
      for (var pro in val) {
        if (val.hasOwnProperty(pro)) { // 这里扩展了对象,所以必须判断
          count++
        }
      }
      // 如果所有的都渲染完了 count可能比实际数量多（重新setSimpleData 数量变少）
      if ((length <= count && length > 0) || params.lazy) {
        params.lockColumnHeight(val)
      }
    })
  }
  computed (params) {
    var that = this
    this.getRowHeight = function (rowIndex) {
      if (params.isLockLeft || params.lazy) {
        return params.lockColumnHeight()[rowIndex] ? params.lockColumnHeight()[rowIndex].height : 'auto'
      } else {
        return 'auto'
      }
    }
    // 状态tab切换或重新加载不同数据后需要清空缓存的列高度
    this.rows.subscribe(function () {
      // 默认getRowHeight是从里面取的所以要把所有的都清空
      params.lockColumnHeight({})
      // 当调用addSimpleData的时候 需要只把原来没变化的设置成needRefresh=true 保留原来的高度（已经渲染过的不会重新走afterrender）
      var allHeight = that.computedLockColumnHeight()
      for (var prop in allHeight) {
        if (allHeight.hasOwnProperty(prop)) { // 这里扩展了对象,所以必须判断
          allHeight[prop].needRefresh = true
        }
      }
      that.computedLockColumnHeight(allHeight)
    }, 'beforeChange')
    this.summaryAfterRender = (elements, data) => {
      // setTimeout(function () {
      ko.tasks.schedule(function () {
        var element = elements.filter(function (el) {
          return el.nodeName === 'TR'
        })[0]
        element.style.width = '100%'
      })
      // })
    }
    this.afterRender = (elements, data) => {
      if (!params.isLockLeft) {
        var lockColumnHeight = that.computedLockColumnHeight()
        if (!lockColumnHeight[data.rowIndex] || lockColumnHeight[data.rowIndex].needRefresh) {
          setTimeout(function () {
            var height = elements.filter(function (el) {
              return el.nodeName === 'TR'
            })[0].offsetHeight
            if (height > 0) {
              lockColumnHeight[data.rowIndex] = {
                height: height + 'px',
                needRefresh: false
              }
              that.computedLockColumnHeight({})
              that.computedLockColumnHeight(lockColumnHeight)
            }
          })
        }
      }
    }
    // td渲染的不同模板
    this.displayMode = (data) => {
      switch (data.col.type) {
        case 'index': return 'y-template-index'
        case 'checkbox': return 'y-template-checkbox'
        case 'render': return 'y-template-render'
        case 'component': return 'y-template-component'
        case 'comp': return 'y-template-component'
        case 'operation': return 'y-template-operation'
        default: return 'y-template-noType'
      }
    }
    // 格式化表格默认值
    this.defaultFormatText = util.defaultFormatText
    this.gridBodyStyle = ko.pureComputed(() => {
      let style = {}
      if (this.lockhead) {
        style.marginTop = this.headHeight
      }
      return style
    })
    // 计算暂无数据的单元格合并
    this.nodataColSpan = ko.computed(() => {
      if (params.columns && params.columns.subscribe) {
        return params.columns().length
      } else {
        return params.columns.length
      }
    })
    // 是否支持多选
    this.isMultiSelect = ko.computed(() => {
      if (params.columns.subscribe) {
        return params.columns().filter(col => {
          return col.type === 'checkbox'
        }).length > 0
      } else {
        return params.columns.filter(col => {
          return col.type === 'checkbox'
        }).length > 0
      }
    })
    this.expandColspan = ko.computed(() => {
      let count = 0
      this.columns().forEach(col => {
        if (ko.utils.unwrapObservable(col._show)) {
          count++
        }
      })
      return count
    })
    // 暂无数据区域的高度
    this.noDataHeight = ko.computed(() => {
      if (params.maxheight === 'auto') {
        return 'auto'
      } else {
        return params.maxheight.replace('px', '') - 150 + 'px'
      }
    })
  }
  created (params) {
    if (params.rowspan) {
      setTimeout(() => {
        ko.tasks.schedule(function () {
          if (params.rowspan.columnIndex) {
            this.caculateRowspanByFields(params.rowspan.columnIndex, params.domId)
          } else if (params.rowspan.maxCol) {
            this.caculateRowspan(params.rowspan.maxCol, params.domId)
          }
        }.bind(this))
      })
      params.rows.subscribe(val => {
        // 确保子组件重绘之后再进行合并单元格，component调用的是后会遇到此问题
        setTimeout(function () {
          ko.tasks.schedule(function () {
            if (params.rowspan.columnIndex) {
              this.caculateRowspanByFields(params.rowspan.columnIndex, params.domId)
            } else if (params.rowspan.maxCol) {
              this.caculateRowspan(params.rowspan.maxCol, params.domId)
            }
          }.bind(this))
        }.bind(this))
      })
    }
  }
  // 判断点击事件是否触发行选中
  isTriggerRowSelect (evt) {
    // 如果点击的是input则不触发
    if (evt.target.tagName === 'INPUT') {
      return false
    }
    if (evt.target.tagName === 'A') {
      return false
    }
    if (evt.target.classList.contains('y-grid-stoprowselect') || evt.target.classList.contains('y-grid-operation') || evt.target.classList.contains('y-select-single')) {
      return false
    }
    return true
  }
  caculateRowspanByFields (indexes, domId) {
    var tab = document.getElementById(domId)
    // val表示单元格的内容
    let val = ''
    let count = 0
    let start = 0
    indexes.forEach(index => {
      // 不同列的合并开始每次需要重置val值
      val = ''
      count = 1
      // 需要减掉用于显示暂无数据的那一行
      for (var i = 0; i < tab.rows.length - 1; i++) {
        // fixed rowspan 在数据切换之后产生的bug
        if (tab.rows[i].cells[index].style.display === 'none') {
          tab.rows[i].cells[index].style.display = 'table-cell'
        }
        // 每次初始化之前要把当前rowSpan重置
        tab.rows[i].cells[index].rowSpan = 1
        // 第一列不需要校验左侧是否被合并了（display === 'none'表示被合并了）
        if (val === tab.rows[i].cells[index].innerHTML) {
          count++
        } else {
          if (count > 1) { // 合并
            start = i - count
            tab.rows[start].cells[index].rowSpan = count
            for (let j = start + 1; j < i; j++) {
              // 被合并的列隐藏
              tab.rows[j].cells[index].style.display = 'none'
            }
            count = 1
          }
          val = tab.rows[i].cells[index].innerHTML
        }
      }
      if (count > 1) { // 合并，最后几行相同的情况下
        start = i - count
        tab.rows[start].cells[index].rowSpan = count
        for (let j = start + 1; j < i; j++) {
          tab.rows[j].cells[index].style.display = 'none'
        }
      }
    })
  }
  // 动态计算rowspan
  caculateRowspan (maxColumns, domId) {
    var tab = document.getElementById(domId)
    var maxCol = maxColumns
    let val = ''
    let count = 0
    let start = 0
    // fix by songhlc 从左边第一列开始判断
    for (var col = 0; col <= maxCol - 1; col++) {
      count = 1
      val = ''
      // 需要减掉用于显示暂无数据的那一行
      for (var i = 0; i < tab.rows.length - 1; i++) {
        // fixed rowspan 在数据切换之后产生的bug
        if (tab.rows[i].cells[col]) {
          if (tab.rows[i].cells[col].style.display === 'none') {
            tab.rows[i].cells[col].style.display = 'table-cell'
          }
          // 每次初始化之前要把当前rowSpan重置
          tab.rows[i].cells[col].rowSpan = 1
          // 第一列不需要校验左侧是否被合并了（display === 'none'表示被合并了）
          if (val === tab.rows[i].cells[col].innerHTML && (col === 0 || tab.rows[i].cells[col - 1].style.display === 'none')) {
            count++
          } else {
            if (count > 1) { // 合并
              start = i - count
              tab.rows[start].cells[col].rowSpan = count
              for (let j = start + 1; j < i; j++) {
                // 被合并的列隐藏
                tab.rows[j].cells[col].style.display = 'none'
              }
              count = 1
            }
            val = tab.rows[i].cells[col].innerHTML
          }
        }
      }
      if (count > 1) { // 合并，最后几行相同的情况下
        start = i - count
        tab.rows[start].cells[col].rowSpan = count
        for (let j = start + 1; j < i; j++) {
          tab.rows[j].cells[col].style.display = 'none'
        }
      }
    }
  }
  methods (params) {
    this.handleMouseOut = (row) => {
      row._hover(false)
      return true
    }
    // mousein
    this.handleMouseIn = (row) => {
      if (row._hover()) return true
      row._hover(true)
      return true
    }
    this.handleClick = (row, evt) => {
      // 如果组织行选中则直接返回不选中任何行
      if (this.forbitRowSelect() || this.forbitRowSelectFn(row)) {
        return false
      }
      // 对于一些输入框不选中
      if (!this.isTriggerRowSelect(evt)) {
        return true
      }
      // todo:需要将选中的数据放入缓存当中或从缓存中剔除
      if (this.isMultiSelect()) {
        let _selected = !row._selected()
        row._selected(_selected)
      } else {
        // 一些单行选中的场景
        this.onRowSelect && this.onRowSelect(row)
      }
      // return true
    }
  }
}
export default Body
