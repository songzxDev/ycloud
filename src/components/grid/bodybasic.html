<div class="y-grid-body" data-bind="style:gridBodyStyle,css:{'y-grid-body-lock':lockhead, 'y-grid-summary':hasSummaryRow}">
  <table cellspacing="0" cellpadding="0" border="0" data-bind="attr:{id:domId},style:{width:tableWidth,transform:transform},css:{'y-table-separate':isSeparate}">
    <colgroup data-bind="foreach:{data: columns,as: 'col'}">
      <col data-bind="visible: col._show, attr:{width:col.width}">
    </colgroup>
    <tbody data-bind="foreach: {data: rows, as: 'row'}, css:{'y-grid-body-verticaltop':verticalAlign === 'top'}">
      <!-- ko ifnot: row._delete -->
        <!-- ko if: !$parent.lazy || ($parent.lazy && $index() >= $parent.lazy.lazyStart() && $index() <= $parent.lazy.lazyEnd()) -->
          <!-- ko template: {name:'y-template-tr',
            data:{
              columns: $parent.columns,
              displayMode:$parent.displayMode,
              afterRender: $parent.afterRender,
              forbitRowSelectFn: $parent.forbitRowSelectFn,
              defaultFormatText: $parent.defaultFormatText,
              rowIndex: $index(),
              row: row,
              rows: $parent.rows,
              getRowHeight: $parent.getRowHeight,
              handleClick: $parent.handleClick,
              handleMouseIn: $parent.handleMouseIn,
              handleMouseOut: $parent.handleMouseOut
            },
            afterRender: $parent.afterRender}
          -->
          <!-- /ko -->
        <!-- /ko -->
        <!-- ko if: $parent.lazy && ($index() < $parent.lazy.lazyStart() || $index() > $parent.lazy.lazyEnd()) -->
        <tr data-bind="style: {
              height: $parent.getRowHeight($index())
              }">
            <td>
            </td>
          </tr>
        <!-- /ko -->
      <!-- /ko -->
    </tbody>
    <tbody class="y-grid-empty">
    <tr>
      <td data-bind="attr:{colspan: nodataColSpan},style:{height: noDataHeight, padding: '12px'}">
        暂无数据
      </td>
    </tr>
    </tbody>
  </table>
  <div class="y-loading-letter-ctn" data-bind="visible:isShowLoading">
    <span class="l-1 y-loading-letter">L</span>
    <span class="l-2 y-loading-letter">o</span>
    <span class="l-3 y-loading-letter">a</span>
    <span class="l-4 y-loading-letter">d</span>
    <span class="l-5 y-loading-letter">i</span>
    <span class="l-6 y-loading-letter">n</span>
    <span class="l-7 y-loading-letter">g</span>
    <span class="l-8 y-loading-letter">.</span>
    <span class="l-9 y-loading-letter">.</span>
    <span class="l-10 y-loading-letter">.</span>
  </div>
</div>
