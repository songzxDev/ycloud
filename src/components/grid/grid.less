.y-grid-wrapper{
  position: relative;
  border: 1px solid @border-color;
  border-bottom: 0;
  border-right: 0;
  margin-bottom: 10px;
  &.y-grid-noborder{
    margin-bottom: 0;
    border:none;
  }
  &.y-grid-separate{
    border: none;
    >.y-grid:before{
      height: 0;
    }
    >.y-grid:after{
      width: 0;
    }
  }
}
.y-grid{
  width: 100%;
  height: 100%;
  max-width: 100%;
  overflow: auto;
  color: #495060;
  font-size: @f-size-base;
  background-color: #fff;
  box-sizing: border-box;
  &.y-grid-border th, &.y-grid-border td{
    border-right: 1px solid @border-color;
  }
  .y-grid-fixed{
    overflow: hidden;
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    box-shadow: 2px 0 6px -2px rgba(0, 0, 0, 0.2);
    .y-grid-fixed-body{
      overflow: hidden;
      position: relative;
    }
  }
  .y-grid-fixed-right{
    overflow: hidden;
    position: absolute;
    top: 0;
    right: 0;
    z-index: 1;
    box-shadow: -2px 0 6px -2px rgba(0, 0, 0, 0.2);
    .y-grid-fixed-body{
      overflow: hidden;
      position: relative;
    }
  }
  .y-grid-operation{
    cursor: pointer;
    color: @f-color-primary;
    margin-right: 10px;
    -webkit-user-select:none;
    -moz-user-select:none;
    -ms-user-select:none;
    user-select:none;
    font-size: @f-size-xs;
    padding-right: 10px;
    border-right: solid 1px #d8d8d8;
  }
  .y-grid-operation-more{
    position: absolute;
    display: none;
    margin-left: -20px;
    background: @bg-color;
    padding: 3px 0;
    z-index: 1;
    border-radius: 2px;
    box-shadow:0 0 5px 0 rgba(0,0,0,0.20);
    li{
      padding: 3px 20px;
      cursor: pointer;
      background: @bg-color;
      &:hover{
        background: #fff;
      }
    }
  }
  .y-grid-operation-morectn {
    .fa{
      position: relative;
      right: 10px;
      top: -3px;
      color: @f-color-primary;
      transition: all .2s ease-in-out;
    }
    .y-grid-operation{
      border-right: none;
      padding-right: 0;
    }
    &:hover {
      .y-grid-operation-more{
        display: block;
      }
      .fa{
        top: 2px;
        transform: rotate(-180deg);
      }
    }
  }
  .y-cell .y-grid-operation:last-child{
    border-right: none;
  }
  &:after, &:before {
    content: "";
    position: absolute;
    background-color: @border-color;
  }
  &:before{
    width: 100%;
    height: 1px;
    left: 0;
    bottom: 0;
    z-index: 1;
  }
  &:after {
    width: 1px;
    height: 100%;
    top: 0;
    right: 0;
    z-index: 3;
  }
  table{
    table-layout: fixed;
    width: 100%;
    tbody {
      tr.y-grid-row-hover td, tr.y-grid-row-selected td{
        background-color: #eff6fc !important;
      }
      .y-grid-checkbox{
        position: relative;
        top: 2px;
        transition: opacity .2s ease-in-out;;
      }
    }
    th{
      height: 40px;
      white-space: nowrap;
      overflow: hidden;
      background-color: #f5f5f5;
      font-weight: normal;
    }
    td{
      background-color: #fff;
      transition: background-color .2s ease-in-out;
    }
    th,td{
      min-width: 0;
      height: 44px;
      box-sizing: border-box;
      text-align: left;
      text-overflow: ellipsis;
      vertical-align: middle;
      color: @f-color-dark;
      border-bottom: 1px solid @border-color;
    }
    .y-cell{
      padding-left: 10px;
      padding-right: 10px;
      line-height: 16px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: normal;
      word-break: break-all;
      box-sizing: border-box;
      &.y-cell-checkbox{
        text-overflow: clip;
      }
    }
    .y-grid-head .y-cell{
      line-height: 16px;
    }
    .y-cell-overflow-auto{
      overflow: visible;
    }
  }
}
.y-grid-stripe .y-grid-body tr:nth-child(2n) td{
  background-color: #f8f8f9;
}
.y-grid-stripe .y-grid-body tr td tr:nth-child(2n + 1) td{
  background-color: #fff;
}
.y-grid-stripe .y-grid-body tr td tr:nth-child(2n) td{
  background-color: #f8f8f9;
}
.y-grid-empty{
  display: none;
  td{
    border-bottom:none !important;
    text-align: center !important;
  }
}
.y-grid-body .y-table-separate{
  border-collapse:separate;
  border-spacing:0px 5px;
  > tbody > tr:nth-child(2n + 1) > td {
    border-top: solid 1px #eee;
  }
  > tbody > tr:nth-child(2n + 1) > td:first-child{
    border-left: solid 1px #eee;
  }
  > tbody > tr:nth-child(2n + 1) > td:last-child{
    border-right: solid 1px #eee;
  }
}
.y-grid-body table tbody:not(.y-grid-empty):empty+.y-grid-empty {
  display: table-row-group;
}
.y-grid-body-verticaltop tr td{
  vertical-align: top;
}
.y-grid-head-lock{
  position: absolute;
  z-index: 1;
  overflow: hidden;
}
.y-grid-body-lock{
  /* margin-top: 45px;*/
}
.y-cell-summary{
  border: solid 1px #e5e6e7;
  position: relative;
  top: 5px;min-height: 44px;
  background-color: #f5f5f5;
  padding: 5px 10px;
}
.y-grid-summary tr {
  td:first-child{
    border-left: solid 1px #e5e6e7;
  }
  td:last-child{
    border-right: solid 1px #e5e6e7;
  }
}
@keyframes yloading {
  0% {opacity: 0;}
  100% {opacity: 1;}
}
.y-loading-letter-ctn{
  text-align: center;
  height: 100%;
  position: absolute;
  width: 100%;
  top: 0;
  padding-top: 20%;
  line-height: 100%;
  color: #333;
  background-color: rgba(0,0,0,.1);
}
.y-loading-letter {
  font-size: 14px;
  animation-name: yloading;
  animation-duration: 1.6s;
  animation-iteration-count: infinite;
  letter-spacing: -1px;
}
.l-1 {animation-delay: .48s;}
.l-2 {animation-delay: .6s;}
.l-3 {animation-delay: .72s;}
.l-4 {animation-delay: .84s;}
.l-5 {animation-delay: .96s;}
.l-6 {animation-delay: 1.08s;}
.l-7 {animation-delay: 1.2s;}
.l-8 {animation-delay: 1.32s;}
.l-9 {animation-delay: 1.44s;}
.l-10 {animation-delay: 1.56s;}
