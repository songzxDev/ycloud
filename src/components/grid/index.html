<div>
  <div class="y-grid-wrapper" data-bind="css: {'y-grid-noborder':noborder, 'y-grid-separate': isSeparate},style:style">
    <div class="y-grid" data-bind="event:{scroll:handleScroll},css:{'y-grid-stripe':isStripe,'y-grid-border': isTableBorder},style:{height:computedMaxHeight,minHeight:minheight}">
      <y-grid-head data-bind="visible:!nohead" params="headtransform:headtransform,el:$el,columns1:columns1,columns2:columns2,lockhead:lockhead,columns:columns,allRowChecked:allRowChecked,tableWidth:tableWidth"></y-grid-head>
      <y-grid-body params="isEnableCrossPage:isEnableCrossPage,crossPageRowPrimaryKey:crossPageRowPrimaryKey,crossPageSelectedRows:crossPageSelectedRows,isShowLoading:isShowLoading,forbitRowSelect:forbitRowSelect,isSeparate:isSeparate,tdstyle: tdstyle, lockhead:lockhead,isNoDataCustomTpl: hasMarkup,defaultExpand:defaultExpand,tableWidth:tableWidth,rowspan:rowspan,domId:domId,expand:expand,maxheight:computedMaxHeight, isDataTable:isDataTable, columns:columns, rows:rows, onRowSelect:onRowSelect">
        <!-- ko template: { nodes: $parent.$templateNodes } --><!-- /ko -->
      </y-grid-body>
    </div>
  </div>
  <!-- ko if: pagination -->
  <y-pagination data-bind="visible: isShowPagination" params="onSizeChange:onSizeChange,onPageChage:onPageChange,pageSize: pageSize, totalCount: totalCount, pageIndex: pageIndex"></y-pagination>
  <!-- /ko -->
</div>
