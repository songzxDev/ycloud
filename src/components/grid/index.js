import template from './index.html'
import ko from 'knockout'
import head from './head'
import body from './body'
import Base from '../../core/base'
const PREFIX = 'y-'
ko.components.register(PREFIX + head.name, {
  viewModel: head.init,
  template: head.template
})
ko.components.register(PREFIX + body.name, {
  viewModel: body.init,
  template: body.template
})
const PAGESIZE = 10
const PAGEINDEX = 0
class Grid extends Base {
  initialize (params) {
    this.isDataTable = params.isDataTable || false
    this.columns = ko.computed(() => {
      // 出事化列内置一些属性 _show 用来隐藏和显示列
      if (params.columns && params.columns.subscribe) {
        params.columns().forEach((col) => {
          // checkbox设定特殊宽度
          if (col.type === 'checkbox') {
            col.width = 35
          }
          col._show = ko.observable(!col.hidden)
        })
        return params.columns()
      } else {
        params.columns.forEach((col) => {
          if (col.type === 'checkbox') {
            col.width = 35
          }
          col._show = ko.observable(!col.hidden)
        })
        return params.columns
      }
    })
    this.rows = params.rows
    this.lockhead = params.lockhead || false
    this.domId = params.id
    this.isTableBorder = params.isTableBorder || params.rowspan
    this.rowspan = params.rowspan
    this.maxheight = params.maxheight || '486px'
    this.minheight = params.minheight || 'auto'
    this.isStripe = params.isStripe || false
    this.expand = params.expand || true
    // 是否显示loading图标
    this.isShowLoading = ko.observable(false)
    // 是否隐藏表头
    this.nohead = params.nohead || false
    this.noborder = params.noborder || false
    // expand是否默认展开
    this.defaultExpand = params.defaultExpand || false
    // 是否启用分页组件
    this.pagination = ko.observable(params.pagination || false)
    this.pageSize = params.pageSize || ko.observable(PAGESIZE)
    this.pageIndex = params.pageIndex || ko.observable(PAGEINDEX)
    this.totalCount = params.totalCount || ko.observable(0)
    this.allRowChecked = ko.observable(false)
    this.tdstyle = params.tdstyle || {}
    // event binding
    // 行选中
    this.onRowSelect = params.onRowSelect || function () {}
    this.onPageChange = params.onPageChange
    this.onSizeChange = params.onSizeChange
    this.isSeparate = params.isSeparate || false
    this.style = params.style || {}
    this.forbitRowSelect = ko.observable(params.forbitRowSelect || false)
    // 放在弹框内部的需要动态内部高度
    this.modalBodyModalHeight = params.modalBodyModalHeight || 0
    this.modalBodyExtraHeight = params.modalBodyExtraHeight || '0'
    this.onModalOkValidate = params.onModalOkValidate || function () {}
  }
  computed (params) {
    // 只有表格数据大于10条才显示分页
    this.isShowPagination = ko.computed(() => {
      return params.pagination && this.totalCount() > 10
    })
    // 表格的宽度
    this.tableWidth = ko.computed(() => {
      let cols
      if (this.columns.subscribe) {
        cols = this.columns()
      } else {
        cols = this.columns
      }
      // 如果存在width不为数字的情况，则认为使用外层的容器宽度，即table宽度是100%
      // 只有在所有列都设置了固定宽度的情况下才支持横向滚动条
      const isUseOuterWidth = cols.some(cell => isNaN(cell.width))
      if (isUseOuterWidth) {
        return '100%'
      } else {
        return cols.map(cell => cell.width).reduce((a, b) => a + b, 0)
      }
    })
    // grid下有传入自定义组件
    this.hasMarkup = ko.computed(() => {
      if (this.$templateNodes.length > 0) {
        let nodeList = this.$templateNodes.filter((node) => {
          return node.nodeName !== '#text' && node.nodeName !== '#comment'
        })
        return nodeList.length > 0
      } else {
        return false
      }
    })
  }
  subscribe (params) {
    // 监听全选和反选
    this.allRowChecked.subscribe(isChecked => {
      // 集成datatable
      this.rows().forEach(row => {
        row._selected(isChecked)
      })
    })
  }
  methods (params) {
    // 设置loading图标是否显示
    this.showLoading = (isShow) => {
      this.isShowLoading(isShow)
    }
    // 获取所有选中的行数据
    this.getSelectedRows = () => {
      let selectedRows = []
      this.rows().forEach((item) => {
        if (item._selected && item._selected()) {
          selectedRows.push(item)
        }
      })
      return selectedRows
    }
    // 设置列显示隐藏
    this.setColVisibleByField = (field, visible) => {
      this.columns().forEach((col) => {
        if (col.field === field) {
          col._show(visible)
        }
      })
    }
    // 切换行是否禁止选中
    this.setRowSelectEnable = (isEnable) => {
      this.forbitRowSelect(isEnable)
    }
    this.computeMaxHeight = () => {
      var maxheight = this.maxheight
      // 如果是在modal中，maxheight = 外层height - extraHeight - paginationHeight - 内层padding高度
      if (maxheight === 'auto') {
        return 'auto'
      } else if (this.modalBodyModalHeight) {
        var _modalBodyModalHeight = this.modalBodyModalHeight.toString().replace('px', '')
        var _modalBodyExtraHeight = this.modalBodyExtraHeight.replace('px', '')
        var _paginationHeight = 0
        var _paddingHeight = 33 // padding+border高度
        if (this.pagination()) {
          _paginationHeight = 36
        }
        var result = _modalBodyModalHeight - _modalBodyExtraHeight - _paginationHeight - _paddingHeight
        // 如果超过允许外层的最大高度
        if (maxheight.replace('px', '') - 0 > result) {
          return result + 'px'
        } else {
          return maxheight
        }
      } else {
        return maxheight
      }
    }
  }
  created (params) {
    this.computedMaxHeight = this.computeMaxHeight()
  }
}
export default {
  name: 'grid',
  init: Base.createViewModel(Grid),
  template
}
