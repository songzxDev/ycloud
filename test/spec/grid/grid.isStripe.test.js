import grid from '../../../src/components/grid/index.js'
import $ from 'jquery'
import kest from '../../utils/kest'
import vm from './viewmodel'
describe('y-grid组件', () => {
  kest.init(grid)
  kest.tpl('<y-grid params="isStripe:true, rows:rows, columns:columns"></y-grid>')
  kest.bind(vm)
  it('默认样式测试', () => {
    expect($('body').find('.y-grid-stripe').length >= 0).toEqual(true)
  })
  it('快照', () => {
    expect($('body').get(0)).toMatchSnapshot()
  })
})
