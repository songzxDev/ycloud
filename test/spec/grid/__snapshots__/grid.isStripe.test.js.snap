// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`y-grid组件 快照 1`] = `
<body>
  <y-grid
    params="isStripe:true, rows:rows, columns:columns"
  >
    <div>
      
  
      <div
        class="y-grid-wrapper"
        data-bind="css: {'y-grid-noborder':noborder, 'y-grid-separate': isSeparate},style:style"
      >
        
    
        <div
          class="y-grid y-grid-stripe"
          data-bind="event:{scroll:handleScroll},css:{'y-grid-stripe':isStripe,'y-grid-border': isTableBorder},style:{height:computedMaxHeight,minHeight:minheight}"
          style="height: 486px; min-height: auto;"
        >
          
      
          <y-grid-head
            data-bind="visible:!nohead"
            params="headtransform:headtransform,el:$el,columns1:columns1,columns2:columns2,lockhead:lockhead,columns:columns,allRowChecked:allRowChecked,tableWidth:tableWidth"
          />
          
      
          <y-grid-body
            params="forbitRowSelectFn:forbitRowSelectFn,caculateColumns:caculateColumns,hasChildGrid:hasChildGrid,verticalAlign:verticalAlign,hasSummaryRow:hasSummaryRow,headHeight:headHeight,isEnableCrossPage:isEnableCrossPage,crossPageRowPrimaryKey:crossPageRowPrimaryKey,crossPageSelectedRows:crossPageSelectedRows,isShowLoading:isShowLoading,forbitRowSelect:forbitRowSelect,isSeparate:isSeparate,tdstyle: tdstyle, lockhead:lockhead,isNoDataCustomTpl: hasMarkup,defaultExpand:defaultExpand,tableWidth:tableWidth,rowspan:rowspan,domId:domId,expand:expand,maxheight:computedMaxHeight, isDataTable:isDataTable, columns:columns, rows:rows, onRowSelect:onRowSelect"
          >
            
        
            <!-- ko template: { nodes: $parent.$templateNodes } -->
            <!-- /ko -->
            
      
          </y-grid-body>
          
      
          <!-- ko if:lockcolumn && rows().length &gt; 0-->
          <!-- /ko -->
          
      
          <!-- ko if:lockright && rows().length &gt; 0-->
          <!-- /ko -->
          
    
        </div>
        
  
      </div>
      
  
      <!-- ko if: pagination -->
      <!-- /ko -->
      

    </div>
    

  </y-grid>
</body>
`;
